tests/code_update.py

```diff
@@ -0,0 +1,238 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the Apache 2.0 License.
+import e2e_args
+import infra.ccf
+import infra.proc
+import json
+import logging
+import os
+import subprocess
+import time
+from infra.ccf import NodeNetworkState
+from loguru import logger as LOG
+from shutil import copyfile
+
+OPCODE_MOV_EAX_DWORD = 0xB8
+OPCODE_PUSH_RBP = 0x55
+OPCODE_RET = 0xC3
+
+
+def patch_binary(so_path, function_name):
+    procs = [
+        ["nm", "-D", so_path],
+        ["grep", f" {function_name}$"],
+        ["cut", "-d", " ", "-f1"],
+    ]
+
+    result = infra.proc.ccall_with_pipe(procs)
+    address = int(result, 0x10)
+
+    with open(so_path, "r+b") as f:
+        f.seek(address, 0)
+        func_prologue = f.read(1)
+        # make sure we know what we patch (the function prologue)
+        if (
+            func_prologue[0] == OPCODE_PUSH_RBP
+            or func_prologue[0] == OPCODE_MOV_EAX_DWORD
+        ):
+            LOG.debug("Patching {} at 0x{:08x}".format(so_path, address))
+            f.seek(address, 0)
+            f.write(bytes([OPCODE_RET]))
+        elif func_prologue[0] == OPCODE_RET:
+            LOG.debug(
+                "Not patching {} at 0x{:08x} - ELF already patched".format(
+                    so_path, address
+                )
+            )
+        else:
+            # The function begins with an instruction which is
+            # neither 'push rbp' nor 'ret', something is not right
+            raise ValueError(
+                "Unexpected function prologue for {}".format(function_name)
+            )
+
+
+def get_code_id(lib_path):
+    oed = subprocess.run(
+        [args.oesign, "dump", "-e", lib_path], capture_output=True, check=True
+    )
+    lines = [
+        line
+        for line in oed.stdout.decode().split(os.linesep)
+        if line.startswith("mrenclave=")
+    ]
+
+    return lines[0].split("=")[1]
+
+
+def create_new_code_version(primary):
+    copyfile(f"{args.package}.so", f"{args.patched_file_name}.so")
+    patch_binary(f"{args.patched_file_name}.so", "stub_for_code_signing")
+
+    # sign the patched binary
+    oed = subprocess.run(
+        [
+            args.oesign,
+            "sign",
+            "-e",
+            f"{args.patched_file_name}.so",
+            "-c",
+            args.oeconfpath,
+            "-k",
+            args.oesignkeypath,
+        ],
+        capture_output=True,
+        check=True,
+    )
+
+    return get_code_id(f"{args.patched_file_name}.so.signed")
+
+
+def vote_to_accept(primary, proposal_id):
```

Comment:
Could this be moved to `ccf` or a new `members`/`governance` module so that it can be re-used? Probably in another PR though

Reply:
It will be done, I will open an issue for this